File: Strings/WordReverse.java
public class WordReverse {
    char[] reverse(char[] A, int l, int r) {
        while (l < r) {
            char temp = A[l];
            A[l] = A[r];
            A[r] = temp;
            l++;
            r--;
        }
        return A;
    }

    char[] wordReverse(char[] A) {
        int n = A.length;

        int l = 0;
        int r = 1;

        reverse(A, 0, n - 1);

        while (r < n) {
            if (A[r] == ' ') {
                reverse(A, l, r - 1);
                l = r + 1;
            }
            r++;
        }
        reverse(A, l, n - 1);

        return A;
    }

    public static void main(String[] args) {
        String s = "I am a student";
        WordReverse w = new WordReverse();
        char[] res = w.wordReverse(s.toCharArray());

        System.out.println(res);
    }
}

File: Strings/wordReverse.py
def reverse(A, l, r):
    while l < r:
        A[l], A[r] = A[r], A[l]
        l += 1
        r -= 1
    return A


def wordReverse(A):
    n = len(A)

    l = 0
    r = 1

    reverse(A, 0, n - 1)

    while r < n:
        if A[r] == " ":
            reverse(A, l, r - 1)
            l = r + 1
        r += 1
    reverse(A, l, r - 1)
    return A


def main():
    # I am a student
    A = ["I", " ", "a", "m", " ", "a", " ", "s", "t", "u", "d", "e", "n", "t"]

    # ['s', 't', 'u', 'd', 'e', 'n', 't', ' ', 'a', ' ', 'a', 'm', ' ', 'I']
    res = wordReverse(A)

    # student a am I
    print("".join(res))


main()
